### ISSUE TITLE
Glitch when destroying the component

### ISSUE THREAD

**Issue by HapLifeMan** [pqina/vue-filepond#82]
Hi, I found something weird when the component is destroyed, I am not able to describe that 😅

![Enregistrement de l’écran 2019-04-01 à 19 32 40](https://user-images.githubusercontent.com/1854141/55347489-57c73e80-54b5-11e9-9733-37bbd6e37493.gif)

**Comment by HapLifeMan** [comment]
Oh sorry, the text isn't important, the FilePond instance is going back to the classic file input when it is destroying.

**Comment by rikschennink** [comment]
It removes itself from the DOM, it seems to do this the moment the popup fades out? Can you adjust the code so it removes itself when the popup is hidden?

**Comment by HapLifeMan** [comment]
I think that's not a pretty solution and hard to do (I'm using Vue transitions with `v-if`).

I tried to modify the source code and I get it work by removing `this._pond.destroy()` in the `beforeDestroy()` method (I tried to changed the `beforeDestroy` to `destroyed` but didn't works).

But this solution can have unexpected behaviors, what do you think of that?

**Comment by rikschennink** [comment]
> I think that's not a pretty solution and hard to do (I'm using Vue transitions with v-if).

Why is that not a pretty solution? It seems logical to only unload the component once it has fully exited the view.

> But this solution can have unexpected behaviors, what do you think of that?

This will keep FilePond running and cause memory leaks.

**Comment by HapLifeMan** [comment]
Of course that solution can cause memory leaks...
So I don't know how to resolve this glitch with Vue transitions without changing the lib 😕

**Comment by rikschennink** [comment]
No way to trigger the cleanup after the transition has ended?

**Comment by HapLifeMan** [comment]
No planned way with a normal use of Vue transitions...

**Comment by rikschennink** [comment]
I've tried to find a solution to this but no luck just yet. It seems the `beforeDestroy` and `destroyed` lifecycle hooks are called before the transition on the parent container has fully completed.

**Comment by AndroAmater** [comment]
Hey guys, It works for me when I change `beforeDestroy` to `destroy` I think your problem was using `destroyed` which is not a Vue lifecycle hook. I'll submit a PR, you can test it and if it solves the problem on your side please merge it :)

**Comment by rikschennink** [comment]
@AndroAmater Appreciate your input, unfortunately this does not solve the issue. `destroy` is not a lifecycle hook (it's not in the diagram) and therefor is not called by Vue, and thus the problem does not occur / seems fixed. `destroyed` is a lifecycle hook, see link below:
https://vuejs.org/v2/api/#destroyed

![Vue Lifecycle Diagram](https://vuejs.org/images/lifecycle.png)
)

**Comment by AndroAmater** [comment]
Shit, my bad. How about `deactivated`? https://vuejs.org/v2/api/#deactivated

**Comment by rikschennink** [comment]
You're free to try, but it seems the description doesn't really match the use case.

**Comment by AndroAmater** [comment]
Well my bad yet again. I was busy so I didn't test. I however did some testing with console logging and found a solution that works but it requires a mixin I found at https://www.npmjs.com/package/vue-removed-hook-mixin

**Comment by rikschennink** [comment]
@AndroAmater Okay, so basically we could add this package as a dependency and then add it to the component which adds the `removed` hook which we then can use to unload the FilePond instance.

I just took a quick look at the code to see how big it was, it's just a couple hundred bytes, which is nice. BUT this is the code.

```js
// adds a 'removed' hook to the component, which fires once the component
// is fully removed from the DOM
export default {
  destroyed() {
    const timer = setInterval(() => {
      if (document.body.contains(this.$el)) {
        // element not yet removed from component
        return
      }

      // quick and dirty version of Vue's lifecycle callHook method
      this.$options.removed.call(this)

      clearInterval(timer)
    }, 25)
  },
}
```

I'mt not super psyched about `setInterval` polling the DOM. We could look into skipping this package and simply observing the DOM in the `destroyed` hook using `MutationObserver`.

**Comment by AndroAmater** [comment]
Well as per your suggestion I played around with `Mutation observer and came up with this`

**Comment by AndroAmater** [comment]
Well as per your suggestion I played around with `MutationObserver` and came up with this
`if (!document.VueFilePondObserver) document.VueFilePondObserver = [];

            document.VueFilePondObserver[this._uid] = new MutationObserver((mutationList, observer) => {
                for(let i = 0; mutationList.length > i; i++) {
                    if(mutationList[i].type === "childList" && mutationList[i].target == this.$el) {

                        if (this._pond) {
                            // bye bye pond
                            this._pond.destroy();
                
                            // remove from instances
                            const index = instances.indexOf(this._pond);
                            if (index >= 0) {
                                instances.splice(index, 1);
                            }
                        }
            

                        if (document.VueFilePondObserver.length <= 1) {
                            document.VueFilePondObserver = undefined;
                        } else {
                            document.VueFilePondObserver[this._uid].disconnect() 
                        }
                    }
                }
            })

            document.VueFilePondObserver[this._uid].observe(this.$parent.$el, {childList: true, subtree: true})`

I tried attaching the `MutationObserver` directly to the element scope but as soon as the element gets destroyed the observer and callback does as well so I ended up attaching it to the document object and created an array so that multiple instances of the component can be used without fighting with each other. I observe the element's parent because it was the only way I could think of to get the outer 'filepond--wrapper' element changes when the component was destroyed. It works for me but I've been staring at the code for a long time now so I would appreciate any suggestions or thoughts on how to make this better, or if you think it would be okay.

**Comment by rikschennink** [comment]
Thanks @AndroAmater I'll look into this and run some tests myself

**Comment by rikschennink** [comment]
Just published version `5.1.1` which fixes the issue in my tests. 

This is the new `destroyed` method.

- It will test if the element has already been detached using `offsetParent`, this is the case when no transitions are used.

- If it is still in the DOM it will observe mutations to the `childList` and will trigger `detached` when it observes the root element has been detached from the DOm.

```js
// Will clean up FilePond instance when unmounted
destroyed () {

    // reference to detached method
    const { detached } = this.$options;
    
    // no longer attached, clean up
    if (!this.$el.offsetParent) {
        detached.call(this);
        return;
    }

    // if we're still attached it's likely a transition is running, we need to 
    // determine the moment when we're no longer attached to the DOM so we can 
    // clean up properly
    const mutationHandler = (mutations, observer) => {
        const removedNodes = (mutations[0] || {}).removedNodes || [];
        if (!removedNodes[0] === this.$el) return;
        observer.disconnect();
        detached.call(this);
    }

    // start observing parent element for changes to the DOM
    const observer = new MutationObserver(mutationHandler);
    observer.observe(this.$parent.$el, { childList: true })
}
```

**Comment by HapLifeMan** [comment]
It's working very well, thank you so much 😊
